### PR TITLE
issue 16393 correct repository name restriction

### LINKED ISSUE #16393: Upgrading docker then unable to run old images due to repository name restrictions
Hi,

See we had an image named like this awesome.com/project/abc_12def3_4_5_6_-_gh-cuont-lol

Old docker client was at 1.6 (even used when at 1.3), suddenly upgraded to 1.8, because why not right, its not a major version change? Right guys?

And now:
repository name component must match "[a-z0-9]+(?:[._-][a-z0-9]+)*"

Its the _-_ which doesnt match that regexp. Which sucks, because why should all our old images suddenly be considered invalid?

### PR DESCRIPTION
**Pull request by iskradelta**
Here is the patch, please include this.
Fix #16393

### CODE REVIEW

File: vendor/src/github.com/docker/distribution/registry/api/v2/names.go [hunk 1]
```diff
@@ -18,7 +18,7 @@ const (
 // RepositoryNameComponentRegexp restricts registry path component names to
 // start with at least one letter or number, with following parts able to
 // be separated by one period, dash or underscore.
-var RepositoryNameComponentRegexp = regexp.MustCompile(`[a-z0-9]+(?:[._-][a-z0-9]+)*`)
+var RepositoryNameComponentRegexp = regexp.MustCompile(`[a-z0-9]+(?:[._-]+[a-z0-9]+)*`)
```
Comment (phemmer) [inline vendor/src/github.com/docker/distribution/registry/api/v2/names.go]: This new regex can be simplified into `[a-z0-9]+[a-z0-9._-]*`

But the comment on the preceeding line (20) now needs to be adjusted, as it's no longer "one period, dash or underscore".

But this is just commentary on the code itself. I do not know if this change is truly acceptable. There must be a reason only one was allowed previously.

Comment (phemmer) [inline vendor/src/github.com/docker/distribution/registry/api/v2/names.go]: Actually no. Regex can't be simplified to that. There is one very slight difference. The regex before ensures that an alnum must follow the `._-`. New one doesn't.

Comment (albers) [inline vendor/src/github.com/docker/distribution/registry/api/v2/names.go]: The pattern now conflicts with the comment.

**Comment by LK4D4** [comment]
Also unittests failing.
ping @tiborvass maybe you know what's going on.

**Comment by iskradelta** [comment]
I think its dumb that the unit test is failing because it checks for
'docker/docker' or docker.io related names as invalid repository name. The
regex cant catch that, its something else which was passing that test.
Which makes the regex not the single point of truth of whats allowed names
and what isnt. So all the current comments are/were wrong too. Please help
me fix this. Or at least lets document that since v2 of registry client its
not backwards compatible with v1 registries hence many images and
dockerfiles are invalid

**Comment by aaronlehmann** [comment]
This file shouldn't be updated directly through a docker/docker PR. It's a vendored copy of a package from the distribution repository.

I'm not sure offhand why this restriction on repeated periods, dashes, and underscores exists.

cc @stevvooe @dmcgowan

**Comment by calavera** [comment]
Please, open this PR in https://github.com/docker/distribution. As @aaronlehmann mentions, this should not be patched here directly.

**Comment by iskradelta** [comment]
@calavera creating a new Issue, the source code of repository should  not be in this repository, since its not patchable here.

**Comment by calavera** [comment]
@iskradelta that's the way go vendoring works. That is not an issue. Please don't create a new issue.

**Pull request closed without merging**